pdns/syncres.cc

```diff
@@ -4611,6 +4698,85 @@ bool SyncRes::processRecords(const std::string& prefix, const DNSName& qname, co
   return done;
 }
 
+
+static void submitTryDotTask(ComboAddress address, const DNSName& auth, const DNSName nsname, time_t now)
+{
+  if (address.getPort() == 853) {
+    return;
+  }
+  address.setPort(853);
+  auto lock = s_dotMap.lock();
+  if (lock->d_numBusy >= SyncRes::s_max_busy_dot_probes) {
+    return;
+  }
+  auto it = lock->d_map.emplace(DoTStatus{address, auth, now + dotFailWait}).first;
+  if (it->d_status == DoTStatus::Busy) {
+    return;
+  }
+  if (it->d_ttd > now) {
+    if (it->d_status == DoTStatus::Bad) {
+      return;
+    }
+    if (it->d_status == DoTStatus::Good) {
+      return;
+    }
+    // We only want to probe auths that we have seen before, auth that only come around once are not interesting
+    if (it->d_status == DoTStatus::Unknown && it->d_count == 0) {
+      return;
+    }
+  }
+  lock->d_map.modify(it, [=] (DoTStatus& st){ st.d_ttd = now + dotFailWait; });
+  bool pushed = pushTryDoTTask(auth, QType::SOA, address, std::numeric_limits<time_t>::max(), nsname);
+  if (pushed) {
+    it->d_status = DoTStatus::Busy;
+    ++lock->d_numBusy;
+  }
+}
+
+static bool shouldDoDoT(ComboAddress address, time_t now)
+{
+  address.setPort(853);
+  auto lock = s_dotMap.lock();
+  auto it = lock->d_map.find(address);
+  if (it == lock->d_map.end()) {
+    return false;
+  }
+  it->d_count++;
+  if (it->d_status == DoTStatus::Good && it->d_ttd > now) {
+    return true;
+  }
+  return false;
+}
+
+static void updateDoTStatus(ComboAddress address, DoTStatus::Status status, time_t time, bool updateBusy = false)
+{
+  address.setPort(853);
+  auto lock = s_dotMap.lock();
+  auto it = lock->d_map.find(address);
+  if (it != lock->d_map.end()) {
+    it->d_status = status;
+    lock->d_map.modify(it, [=] (DoTStatus& st) { st.d_ttd = time; });
+    if (updateBusy) {
+      --lock->d_numBusy;
+    }
+  }
+}
+
+bool SyncRes::tryDoT(const DNSName& qname, const QType qtype, const DNSName& nsName, ComboAddress address, time_t now)
+{
+  LWResult lwr;
+  bool truncated;
+  bool spoofed;
+  boost::optional<Netmask> nm;
+  address.setPort(853);
+  // We use the fact that qname equals auth
+  bool ok = doResolveAtThisIP("", qname, qtype, lwr, nm, qname, false, false, nsName, address, true, true, truncated, spoofed);
```

Comment:
It looks like we will not update the status of that entry if `doResolveAtThisIP` throws, and that would mean that we keep an entry in a `Busy` state, still counted in `d_numBusy` as it should, but we do not have a task for it anymore?